### DISCUSSION TITLE
Understanding CMakePresets.json inheritance rules

### DISCUSSION THREAD

**Discussion by daldr-ntml** [microsoft/vscode-cmake-tools#2572] [Q&A] [answered]
Hi, I'm not sure I understand the rules of inheritance for CMakePresets.json. Here is part of my json file:

```
{
  "version": 2,

  "configurePresets": [
    {
      "name": "linux-remote-base",
      "description": "Base configuration for remote Ubuntu targets, using Ninja generator.",
      "hidden": true,
      "generator": "Ninja",
      "binaryDir": "${sourceDir}/builds/${presetName}",
      "cacheVariables": {
      },
      "environment": {
        "MKLROOT": "/opt/intel/oneapi/mkl/2021.1.1",
      },
      "vendor": {
        "microsoft.com/VisualStudioSettings/CMake/1.0": { "hostOS": [ "Linux" ] },
        "microsoft.com/VisualStudioRemoteSettings/CMake/1.0": {
          "sourceDir": "<snip>",
          "copyExcludeDirectoriesList": [ "/opt/rh/devtoolset-9", "/opt/dpdk/dpdk-stable-18.11.9", "/opt/intel/oneapi/mkl/latest" ],
          "copySourcesOptions": {
            "method": "rsync",
            "concurrentCopies": 10,
            "outputVerbosity": "Normal"
          }
        }
      }
    },
    {
      "name": "linux-all-base",
      "description": "Base configuration to target Linux",
      "hidden": true,
      "inherits": "linux-remote-base",
      "cacheVariables": {
        "PROJECT_VARIANT": "MyVariant", 
      },
      "vendor": {
        "microsoft.com/VisualStudioRemoteSettings/CMake/1.0": {
          "rsyncCommandArgs": [ "-tr  --exclude-from=rsync_exclude "],
          "copySourcesOptions": {
            "exclusionList": [ "*.tmp" ]
          }
        }
      }
    },
    {
      "name": "ubuntu-all-release",
      "description": "Target an Ubuntu system, Release build.",
      "hidden": false,
      "inherits": "linux-all-base",
      "cacheVariables": {
        "CMAKE_BUILD_TYPE": "Release"
      },
      "vendor": {}
    }
  ]
```
There are 3 configuration presets, each inheriting from the one immediately above. My questions are for the net effect in configuration "ubuntu-all-release" (the top of the hierarchy).

Q1 Will _cacheVariables_ include variable 'PROJECT_VARIANT' or just 'CMAKE_BUILD_TYPE'? (i.e. does the top _cacheVariables_ override all settings of the lower ones?)

Q2 In the vendor settings is setting them to {} in this top config, overriding (i.e. removing) all the vendor settings below?

Q3 If not, will vendor settings contain both _copySourcesOptions.method_ and _copySourcesOptions.exclusionList_ or only the latter?

Q4 Is there a way of 'seeing' the net inherited values for verification?

**Comment by bobbrow** [comment, marked as answer]
Q1 - Values are overridden at the lower level, not at the object level (at least for `cacheVariables` and `environment`), so having different variables defined inside `cacheVariables` should result in a merged `cacheVariables with both `PROJECT_VARIANT` and `CMAKE_BUILD_TYPE` defined.

Q2 - I would expect the same rules from Q1 to apply here, so for the example provided, I would not expect anything to be removed from the vendor settings because omitting a property is not deleting it. To delete settings you set them to `null`, though I don't know how well this was tested across all the properties and it's possible that there could be bugs here.  I need to dig into this a bit more to see what Visual Studio decided here.

Q3 - Similar to Q2, I need to see what Visual Studio decided for the remote settings (I know who to talk to - my team didn't implement this part)

Q4 - Not currently.  I thought we had an issue tracking this specifically, but I think #1945 is the closest we came to saying that.  I would appreciate if you could open a feature request issue for this.

**Reply by daldr-ntml** [reply]
@bobbrow Thank you for your answer. I have opened https://github.com/microsoft/vscode-cmake-tools/issues/2575.
